### PR TITLE
[AD-372] Fix incorrect join output for nested array and nested document in array

### PR DESCRIPTION
**Pull request by andiemontoyeah**
### Summary

-  Fix incorrect join output for nested array and nested document in array

### Description

- $unwind could not be after $addFields for this particular case because the field we want to add is the index created from $unwind.  
- Added logic to flag when $addFields relies on $unwind.  
- Reorder $addFields and $unwind as a very last step to make sure output is still correct for multiple tables.

### Related Issue
https://bitquill.atlassian.net/browse/AD-372

### Additional Reviewers
@affonsoBQ
@andiem-bq
@alexey-temnikov
@birschick-bq
@mitchell-elholm

### CODE REVIEW

File: src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementJoinTest.java [hunk 1]
```diff
@@ -1047,4 +1047,110 @@ void testCrossJoin(final DocumentDbTestEnvironment testEnvironment) throws SQLEx
...
+        try (Connection connection = getConnection()) {
+            final DocumentDbStatement statement = getDocumentDbStatement(connection);
+
+            // Verify LEFT OUTER JOIN on the nested table and 2nd nested table to produce 2 rows.
+            final ResultSet resultSet1 =
+                    statement.executeQuery(
+                            String.format(
+                                    "SELECT * FROM \"%s\".\"%s\" "
+                                            + "LEFT OUTER JOIN \"%s\".\"%s\" "
+                                            + "ON \"%s\".\"%s\" = \"%s\".\"%s\" "
```
Comment (birschick-bq) [inline src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementJoinTest.java]: nit: Could use positional place holders (i.e., `%1$s`, `%2$s`, etc.) so you don't have to repeat the same inputs multiple times.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbJoin.java [hunk 2]
```diff
@@ -195,46 +195,76 @@ private void joinSameCollection(
...
+                                    .build();
                     columnMap.put(newKey, newRightColumn);
 
-                    // Handle any column renames
-                    handleColumnRename(renames, newKey, entry.getValue().getFieldPath(),
-                            rightIsVirtual, rightFilterColumns);
-                    handleColumnRename(renames, leftColumn.getFieldPath(), leftColumn.getFieldPath(),
-                            leftIsVirtual, leftFilterColumns);
+                    if (column.isIndex()) {
+                        resolutionNeedsUnwind = true;
```
Comment (birschick-bq) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbJoin.java]: nit: `resolutionNeedsUnwind = column.isIndex();`

File: src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementJoinTest.java (hunk 1 above)
Comment (andiemontoyeah) [reply src/test/java/software/amazon/documentdb/jdbc/DocumentDbStatementJoinTest.java]: ty!! helps a lot with keeping <2000 lines too :D

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbRules.java [hunk 3]
```diff
@@ -148,12 +148,12 @@ static String isItem(final RexCall call) {
 
     static String getPath(final DocumentDbSchemaColumn column, final boolean useOriginalPaths) {
         final String path;
-        if (column.isIndex()) {
-            path = column.getSqlName();
-        } else if (column instanceof DocumentDbMetadataColumn
+        if (column instanceof DocumentDbMetadataColumn
```
Comment (andiemontoyeah) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbRules.java:151]: Reordered if/else ... path for indexes that were projected with a rename was not correct

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbTableScan.java [hunk 4]
```diff
@@ -124,8 +124,15 @@ protected DocumentDbTableScan(final RelOptCluster cluster, final RelTraitSet tra
                 arrayPath = "$" + arrayPath;
```
Comment (andiemontoyeah) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbTableScan.java:124]: Determine any needed $unwind and virtual table null filter at the beginning but do not add until last step.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbToEnumerableConverter.java [hunk 5]
```diff
@@ -158,19 +158,29 @@ protected DocumentDbToEnumerableConverter(
     }
 
     /**
-     * Adds aggregation stage to handle virtual tables that may be null.
+     * Adds aggregation stage to handle virtual tables.
      * @param implementor the implementor.
      */
     public static void handleVirtualTable(final Implementor implementor) {
```
Comment (andiemontoyeah) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbToEnumerableConverter.java:164]: Insert here the needed $addFields, $unwinds and virtual table filters. 
When $addFields is using an index, we need to $unwind first. In other cases, $addFields can be first to enhance performance.

**Review by birschick-bq — APPROVED** [review]
I think this LGTM.

**Review by alexey-temnikov — APPROVED** [review]

**Pull request merged**